[BOLT] Overhaul the comments in PAuthGadgetScanner for readability (NFC)

atrosinenko opened the pull request:
Update the comments in PAuthGadgetScanner.cpp to better describe the
current version of the code. Along the way, shorten identifier names
that are redundant taking their context into account:
`RegsToTrackInstsFor` (made `RegsToTrack`) and `getNumTrackedRegisters`
(made `getNumRegisters`).

kbeyls reviewed (approved):
Thank you very much for improving the documentation like this!

I like it overall, I only left a few minor nit-picky suggestions; feel free to disagree on all of them.

One thing that crossed my mind by the time I read through the whole patch, is that the different classes and how they interact continues to be quite complicated.
I wonder if the long comment at the top of the file could be extended with maybe something akin to a UML diagram or something similar to show what each class is exactly responsible for and how they interact.
It's just a thought though. I'm not sure at all if such a comment would improve things.
I thought I'd just mention it here in case you think it would improve being able to understand better how the different parts of the code all stitch together.

atrosinenko commented on bolt/lib/Passes/PAuthGadgetScanner.cpp:
```diff
@@ -6,8 +6,40 @@
-// that may not be protected by PAuth authentication instructions when needed.
+// This file implements a pass that looks for instructions which are expected
+// to be protected by Pointer Authentication, but the protection is missing
+// or insufficient. While the existing implementation only applies to AArch64,
```
Looks like I confused the intention and the implementation. My idea was that there are specific instructions whose inputs or outputs should be protected, but this looks more like an implementation detail of this scanner, and not the original problem statement. I would say "instructions which are expected to be present" is kind of implementation detail, too. So far, I updated this to be

> This file implements a pass that analyzes code hardened using Pointer
> Authentication and looks for non-protected or insufficiently protected parts.

Maybe "... insufficiently protected control flow" would sound better, but PtrAuth is not necessarily restricted to control flow...

atrosinenko commented on bolt/lib/Passes/PAuthGadgetScanner.cpp:
```diff
@@ -162,14 +164,20 @@ typedef SmallPtrSet<const MCInst *, 4> SetOfRelatedInsts;
+  /// instructions that perform memory access and generate an error on failed
+  /// address translation. These registers are not generally safe to be used
+  /// by the instructions like pointer signing, as such usage may hide the
+  /// authentication failure.
```
This is probably not correct: it is usually safe to perform address arithmetic on such values, for example, provided its result is used in a safe manner. Or it is *possibly* safe to feed it as yet another source of entropy to a PRNG :) The problem of re-signing an unchecked pointer is that depending on the specific semantic of the sign operation (which probably depends on features like `FEAT_PAuth2` on AArch64), it may produce a value indistinguishably from a perfectly valid signed pointer that we have to trust after successful authentication.

atrosinenko reviewed (commented):
Thank you for the comments, addressed them in ec8d7e83b5637689d0c888e5e14fbfb2b2b7a1af, please feel free to request further refinements.

atrosinenko committed:
Address the review comments

Co-authored-by: Kristof Beyls <kristof.beyls@arm.com>

atrosinenko committed:
[BOLT] Overhaul the comments in PAuthGadgetScanner for readability (NFC)

Update the comments in PAuthGadgetScanner.cpp to better describe the
current version of the code. Along the way, shorten identifier names
that are redundant taking their context into account:
`RegsToTrackInstsFor` (made `RegsToTrack`) and `getNumTrackedRegisters`
(made `getNumRegisters`).
